engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,303 @@
+                            trace!("{}", e);
+                        } else {
+                            message_request.completed = true; // Mark as completed
+                        }
+                    }
+                } else {
+                    panic!("Event not found: {}", message_request.name);
+                }
+            }
+
+            for message_request in message_requests {
+                if message_request.stay_alive || !message_request.completed {
+                    message_heap.push(message_request);
+                }
+            }
+        }
+    }
+
+    pub fn dispatch_all(&mut self, mut lua: Ref<Lua>) {
+        for update_pass in UpdatePass::iter() {
+            self.dispatch(update_pass, lua.borrow_mut());
+        }
+    }
+}
+
+#[luajit_ffi_gen::luajit_ffi]
+impl EventBus {
+    pub fn register(
+        &mut self,
+        event_name: String,
+        priority: Option<u16>, //* how do i handle Options via ffi? It requires a uint16 const pointer */
+        update_pass: UpdatePass, //* how do i make the update pass enum work nicely for the lua side? Since the enum is a lua number on lua side.*/
+        with_update_pass_message: bool,
+    ) {
+        let priority = priority.unwrap_or(0);
+
+        if self.max_priority_locked {
+            if priority == u16::MAX {
+                panic!("Trying to register event at locked priority");
+            }
+        }
+
+        let event = Event {
+            name: event_name.clone(),
+            priority,
+            update_pass: update_pass,
+            subscribers: vec![],
+            // payloads,
+        };
+
+        match self.events.entry(event_name.clone()) {
+            Entry::Occupied(_) => {
+                println!(
+                    "Warning: You are trying to register an Event '{}' that already exists - Aborting!",
+                    event_name
+                );
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(event);
+
+                if with_update_pass_message {
+                    let event_item = MessageRequest {
+                        priority,
+                        name: event_name.clone(),
+                        stay_alive: with_update_pass_message,
+                        completed: false,
+                    };
+
+                    self.update_pass_map
+                        .entry(update_pass)
+                        .or_insert_with(|| {
+                            println!("Inserting new BinaryHeap for {:?}", update_pass);
+                            BinaryHeap::new()
+                        })
+                        .push(event_item.clone());
+
+                    // Verify the message_heap immediately after insertion
+                    if let Some(message_heap) = self.update_pass_map.get(&update_pass) {
+                        let message_requests: Vec<_> = message_heap.clone().into_sorted_vec();
```

Comment:
The heap is already sorted, and we're cloning the _entire_ message heap here. Perhaps just return an iterator over the heap? i.e. `for message_request in message_heap.iter() { ...`